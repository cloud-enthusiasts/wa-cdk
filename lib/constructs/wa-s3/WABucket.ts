import {BlockPublicAccess, Bucket, BucketEncryption, BucketProps, IBucket} from "@aws-cdk/aws-s3";
import { v4 as uuid } from 'uuid';
import {Construct} from "@aws-cdk/core";

export interface WellArchitectedBucketProps extends BucketProps {

    /*
     * This property is used to enable or disable bucket security, defaults to true
     */
    readonly security?: boolean;

    /*
     * This property is used to enable to disable bucket monitoring, defaults to true
     *
     */
    readonly monitoring?: boolean;

    /*
     * If monitoring is enabled then logBucket name must be provided and it should already exist in the environment.
     */
    readonly logBucketArn?: string;

}

/*
 *  Well Architected S3 Bucket
 */
export class WABucket extends Bucket {

    constructor(scope: Construct, id: string, props: WellArchitectedBucketProps) {

        super(scope, id, {
            ...props,
            bucketName: WABucket.getValueOrDefault(props.bucketName, WABucket.generateRandomizedName(id)),
            encryption: WABucket.shouldEncryptAtRest(props),
            enforceSSL: WABucket.shouldEncryptInTransit(props),
            blockPublicAccess: WABucket.shouldBlockPublicAccess(props),
            serverAccessLogsBucket: WABucket.shouldEnableLogging(scope, props),
        });
    }

    private static generateRandomizedName(id: string): string {
        return `${uuid()}-${id.toLowerCase()}`
    }

    private static shouldEnableLogging(scope: Construct, props: WellArchitectedBucketProps): IBucket | undefined {
        if(WABucket.shouldMonitor(props)) {
            if(props.logBucketArn != undefined) {
                return Bucket.fromBucketArn(scope, "S3Logs", props.logBucketArn)
            } else {
                throw new Error("When monitoring is enabled, valid logBucketArn must be provided.")
            }
        } else {
            return undefined;
        }
    }

    private static shouldEncryptAtRest(props: WellArchitectedBucketProps): BucketEncryption {
        if(WABucket.shouldSecure(props)) {
            return WABucket.getValueOrDefault(props.encryption, BucketEncryption.S3_MANAGED);
        } else {
            return BucketEncryption.UNENCRYPTED;
        }
    }

    private static shouldEncryptInTransit(props: WellArchitectedBucketProps): boolean {
        if(WABucket.shouldSecure(props)) {
            return WABucket.getValueOrDefault(props.enforceSSL, true);
        } else {
            return false;
        }
    }

    private static shouldSecure(props: WellArchitectedBucketProps) {
        return WABucket.getValueOrDefault(props.security, true)
    }

    private static shouldMonitor(props: WellArchitectedBucketProps) {
        return WABucket.getValueOrDefault(props.monitoring, true)
    }

    private static shouldBlockPublicAccess(props: WellArchitectedBucketProps): BlockPublicAccess {
        if(WABucket.shouldSecure(props)) {
            return WABucket.getValueOrDefault(props.blockPublicAccess, BlockPublicAccess.BLOCK_ALL);
        } else {
            return BlockPublicAccess.BLOCK_ALL;
        }
    }

    private static getValueOrDefault<T>(propertyValue: T | undefined, defaultValue: T): T {
        if( propertyValue != null || propertyValue != undefined)
            return propertyValue;
        else {
            return defaultValue;
        }
    }
}