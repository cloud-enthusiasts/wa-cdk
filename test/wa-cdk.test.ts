import {App, Stack} from "@aws-cdk/core";
import {WABucket} from "../lib/constructs/wa-s3/WABucket";
import {expect as cdkExpect, haveResource, SynthUtils} from "@aws-cdk/assert";

let app: App;
let stack: Stack;

describe("Well Architected Bucket Properties Tests", () => {

    beforeEach(() => {
        app = new App();
        stack = new Stack(app, "Test-Stack", {});
    });

    test("Bucket should be secure by default", () => {
        new WABucket(stack, "MyTestBucket", {
            bucketName: "mytestbucket",
            monitoring: false
        });

        cdkExpect(stack).to(haveResource("AWS::S3::Bucket", {
            "PublicAccessBlockConfiguration": {
                "BlockPublicAcls": true,
                "BlockPublicPolicy": true,
                "IgnorePublicAcls": true,
                "RestrictPublicBuckets": true,
            },
            "BucketEncryption": {
                "ServerSideEncryptionConfiguration": [{
                        "ServerSideEncryptionByDefault": {
                            "SSEAlgorithm": "AES256",
                        }
                    }]
            }
        }));

        cdkExpect(stack).to(haveResource("AWS::S3::BucketPolicy", {
            "PolicyDocument":  {
            "Statement":  [
                 {
                "Action": "s3:*",
                    "Condition":  {
                    "Bool":  {
                        "aws:SecureTransport": "false",
                    },
                },
                "Effect": "Deny",
                    "Principal":  {
                    "AWS": "*",
                },
                "Resource":  [
                     {
                    "Fn::GetAtt":  [
                        "MyTestBucket81062429",
                            "Arn",
                        ],
                },
                 {
                    "Fn::Join":  [
                        "",
                             [
                                 {
                        "Fn::GetAtt":  [
                            "MyTestBucket81062429",
                                "Arn",
                            ],
                    },
                    "/*",
                ],
                ],
                },
            ],
            },
        ],
            "Version": "2012-10-17",
        }
        }));
        expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
    });

    it("should enable monitoring by default", () => {
        new WABucket(stack, "MyTestBucket", {
            bucketName: "mytestbucket",
            logBucketArn: "arn:aws:s3:::mytestlogbucket"
        });

        cdkExpect(stack).to(haveResource("AWS::S3::Bucket", {
            "LoggingConfiguration": {
                "DestinationBucketName": "mytestlogbucket",
            }
        }));

        expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
    });

    it("should allow monitoring to be turned off", () => {
        new WABucket(stack, "MyTestBucket", {
            bucketName: "mytestbucket",
            monitoring: false
        });

        expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
    });

})
