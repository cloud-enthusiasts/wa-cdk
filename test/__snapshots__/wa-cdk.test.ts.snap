// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Well Architected Bucket Properties Tests Bucket should be secure by default 1`] = `
Object {
  "Resources": Object {
    "MyTestBucket81062429": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "BucketName": "mytestbucket",
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "MyTestBucketPolicyE11AF29F": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "MyTestBucket81062429",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:*",
              "Condition": Object {
                "Bool": Object {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": Object {
                "AWS": "*",
              },
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "MyTestBucket81062429",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "MyTestBucket81062429",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
  },
}
`;

exports[`Well Architected Bucket Properties Tests should allow monitoring to be turned off 1`] = `
Object {
  "Resources": Object {
    "MyTestBucket81062429": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "BucketName": "mytestbucket",
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "MyTestBucketPolicyE11AF29F": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "MyTestBucket81062429",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:*",
              "Condition": Object {
                "Bool": Object {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": Object {
                "AWS": "*",
              },
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "MyTestBucket81062429",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "MyTestBucket81062429",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
  },
}
`;

exports[`Well Architected Bucket Properties Tests should enable monitoring by default 1`] = `
Object {
  "Resources": Object {
    "MyTestBucket81062429": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "BucketName": "mytestbucket",
        "LoggingConfiguration": Object {
          "DestinationBucketName": "mytestlogbucket",
        },
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "MyTestBucketPolicyE11AF29F": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "MyTestBucket81062429",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:*",
              "Condition": Object {
                "Bool": Object {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": Object {
                "AWS": "*",
              },
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "MyTestBucket81062429",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "MyTestBucket81062429",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
  },
}
`;
